draft-specifications/CapTP Specification.md

```diff
@@ -0,0 +1,978 @@
+    to and from when crossing a CapTP boundary.
+-   [OCapN Netlayers](): Specification to open a secure communication channel
+    between two sessions, often on different networks.
+-   [OCapN Locators](): Specification covers representation of object references
+    for both in band and out of band usage.
+
+# CapTP Overview
+
+CapTP supports many features, as described in the [Introduction](#introduction)
+section. It can, however, be broken down into the life cycle of a CapTP session.
+A session is between two entities which are communicating via CapTP through a
+OCapN Netlayer's channel. A CapTP life cycle consists of the following stages:
+
+1. A session is established pairwise between machines.
+2. Communication occurs between objects.
+    -   Initial connectivity to objects is established by querying the bootstrap
+        object for access to known objects (via sturdyrefs or certificates).
+    -   Messages are exchanged between objects which hold references to each
+        other.
+    -   Promises may be created when messages are sent.
+    -   Messages may be pipelined to promises, queueing those messages to
+        eventually be delivered to their resolution.
+    -   Handoffs are initiated when sending a message with a reference to an
+        object on a third CapTP session.
+    -   Both parties cooperate to free object references which are recognized to
+        no longer be needed to be imported from one side.
+5.  A session ends.
+
+## Establishing a connection
+
+A new connection is established either by:
+
+- Using an out-of-band mechanism to connect to a remote session
+- Peforming a handoff to connect to a third session
+
+In either situation, a secure channel needs to be created to the remote session.
+This is out of scope for this specification but it's covered by the OCapN
+Netlayers specification. Once a secure channel is established, we MUST to
+perform the following in this order:
+
+1.  Create a per-session cryptography key-pair (see
+    [Cryptography](#cryptography))
+2.  Send a [`op:start-session`](#op-start-session) message
+3.  Receive and verify the remote party's
+    [`op:start-session`](#op-start-session) (including dialback verification)
+5.  Send a [`op:bootstrap`](#op-bootstrap) to import their bootstrap object
+
+(**NOTE:** The `op:bootstrap` is likely to be deprecated soon.)
+
+Once these steps are successfully completed, the connection is considered
+established and set up, and can be used to send and receive messages with remote
+objects.
+
+## Sending and receiving messages
+
+When an object sends a message to another object across a CapTP boundary, it
+either expects a reply or not. If the object expects a reply a
+[promise](#promises) is created to represent the reply. To do this, we send a
+[`op:deliver`](#op-deliver) message. If the message is delivered without
+importing an answer, this is an [`op:deliver-only`](#op-deliver-only) message.
```

Comment:
This paragraph uses "a `op:deliver`" and "an `op:deliver-only`". Should these be made consistent?

Reply:
@davexunit, I believe these are two different messages, the latter of which ignores any result.

Reply:
Oh let me be more clear: It's the "a" and "an" part that is inconsistent to me.